[iPad] store.steampowered.com: Navigation bar doesn’t reveal drop down menus when tapped

whsieh opened the pull request:
#### ce2bebc6a7aa4cc6dff7101c531442d2f5b4ac46
<pre>
[iPad] store.steampowered.com: Navigation bar doesn’t reveal drop down menus when tapped
<a href="https://bugs.webkit.org/show_bug.cgi?id=282763">https://bugs.webkit.org/show_bug.cgi?id=282763</a>
<a href="https://rdar.apple.com/89114638">rdar://89114638</a>

Reviewed by Richard Robinson and Alan Baradlay.

When tapping items in the top navigation menu on store.steampowered.com, the page reveals a hover
menu with more links. This currently manages to bypass all existing content change observation
heuristics, leading to a `click` event being dispatched to the page after `hover`.

At the same time, Steam also adds a click event handler that *dismisses* the open menu when handling
mouse hover. The end result is that Steam schedules logic to begin revealing the menu when handling
`mouseover`, and then immediately dismisses the menu when handling the synthetic `click` event ~32
ms later.

To fix this, we make a small adjustment to `ContentChangeObserver`, so that we recognize the tap
gesture on these menus on steampowered.com as *only* a mouse hover, rather than clicking. While
there isn&apos;t any significant visual change in response to the hover event, we can detect that the
page adds a `mouseout` event listener to a container of the click target during the scope of content
change observation, which is a fairly strong signal that the page intends to know when the mouse has
exited the click target.

* LayoutTests/fast/events/touch/ios/content-observation/add-mouseout-event-listener-on-hover-expected.txt: Added.
* LayoutTests/fast/events/touch/ios/content-observation/add-mouseout-event-listener-on-hover.html: Added.

Add a test case to exercise this heuristic.

* Source/WebCore/dom/Document.h:
(WebCore::Document::contentChangeObserverIfExists):
* Source/WebCore/dom/Node.cpp:
(WebCore::tryAddEventListener):
* Source/WebCore/page/ios/ContentChangeObserver.cpp:
(WebCore::ContentChangeObserver::reset):
(WebCore::ContentChangeObserver::didAddMouseMoveRelatedEventListener):
(WebCore::ContentChangeObserver::adjustObservedState):
* Source/WebCore/page/ios/ContentChangeObserver.h:
(WebCore::ContentChangeObserver::setClickTarget):

Store the current click target (`nodeRespondingToClick`) on the content change observer, so that we
can check against it when observing any event handlers being added on the fly during content
observation.

* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::handleSyntheticClick):

Canonical link: <a href="https://commits.webkit.org/286320@main">https://commits.webkit.org/286320@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9226fe5bb24624026da4dae2c737be701b947a1f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/75530 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/192 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/28381 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/80009 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/26794 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/77646 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/192 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/2761 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/59251 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/17448 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/78597 "Passed tests") | [⏳ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/iOS-17-Simulator-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/64884 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/39609 "Passed tests") | 
| | [⏳ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/iOS-17-Simulator-WPT-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/22369 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/25122 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/67812 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/22706 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/81489 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/2869 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/1812 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/67493 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/3020 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/64860 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/66786 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16662 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/10745 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/8901 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/2826 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/5647 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/2851 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/3786 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/2858 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/dom/Node.cpp:
```diff
@@ -2456,6 +2456,14 @@ static inline bool tryAddEventListener(Node* targetNode, const AtomString& event
 #if ENABLE(IOS_GESTURE_EVENTS) && ENABLE(TOUCH_EVENTS)
     if (typeInfo.isInCategory(EventCategory::Gesture))
         document->addTouchEventHandler(*targetNode);
+
```
Nit: Unnecessary newline.

whsieh replied:
Good catch!

whsieh committed:
[iPad] store.steampowered.com: Navigation bar doesn’t reveal drop down menus when tapped
https://bugs.webkit.org/show_bug.cgi?id=282763
rdar://89114638

Reviewed by Richard Robinson and Alan Baradlay.

When tapping items in the top navigation menu on store.steampowered.com, the page reveals a hover
menu with more links. This currently manages to bypass all existing content change observation
heuristics, leading to a `click` event being dispatched to the page after `hover`.

At the same time, Steam also adds a click event handler that *dismisses* the open menu when handling
mouse hover. The end result is that Steam schedules logic to begin revealing the menu when handling
`mouseover`, and then immediately dismisses the menu when handling the synthetic `click` event ~32
ms later.

To fix this, we make a small adjustment to `ContentChangeObserver`, so that we recognize the tap
gesture on these menus on steampowered.com as *only* a mouse hover, rather than clicking. While
there isn't any significant visual change in response to the hover event, we can detect that the
page adds a `mouseout` event listener to a container of the click target during the scope of content
change observation, which is a fairly strong signal that the page intends to know when the mouse has
exited the click target.

* LayoutTests/fast/events/touch/ios/content-observation/add-mouseout-event-listener-on-hover-expected.txt: Added.
* LayoutTests/fast/events/touch/ios/content-observation/add-mouseout-event-listener-on-hover.html: Added.

Add a test case to exercise this heuristic.

* Source/WebCore/dom/Document.h:
(WebCore::Document::contentChangeObserverIfExists):
* Source/WebCore/dom/Node.cpp:
(WebCore::tryAddEventListener):
* Source/WebCore/page/ios/ContentChangeObserver.cpp:
(WebCore::ContentChangeObserver::reset):
(WebCore::ContentChangeObserver::didAddMouseMoveRelatedEventListener):
(WebCore::ContentChangeObserver::adjustObservedState):
* Source/WebCore/page/ios/ContentChangeObserver.h:
(WebCore::ContentChangeObserver::setClickTarget):

Store the current click target (`nodeRespondingToClick`) on the content change observer, so that we
can check against it when observing any event handlers being added on the fly during content
observation.

* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::handleSyntheticClick):

Canonical link: https://commits.webkit.org/286320@main
